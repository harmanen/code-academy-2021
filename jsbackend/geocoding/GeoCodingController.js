import axios from "axios";
import querystring from "querystring";

const { HERE_API_KEY, MAPBOX_API_KEY, OPENCAGE_API_KEY } = process.env;

export const getMapBoxLocation = async (address, axios) => {
  const response = await axios.get(
    `https://api.mapbox.com/geocoding/v5/mapbox.places/${address}.json?` +
      querystring.stringify({ access_token: MAPBOX_API_KEY })
  );
  const [lng, lat] = response.data?.features?.[0]?.center || [null, null];
  return { lng, lat };
};

export const getHereLocation = async (address, axios) => {
  const response = await axios.get(
    `https://geocode.search.hereapi.com/v1/geocode?` +
      querystring.stringify({ q: address, apiKey: HERE_API_KEY })
  );
  return response.data?.items?.[0].position || { lat: null, lng: null };
};

export const getOpenCageLocation = async (address, axios) => {
  const response = await axios.get(
    `https://api.opencagedata.com/geocode/v1/json?` +
      querystring.stringify({ q: address, key: OPENCAGE_API_KEY })
  );
  return response.data?.results?.[0]?.geometry || { lat: null, lng: null };
};

export const ALL_GEOCODE_FUNCTIONS = [
  getMapBoxLocation,
  getHereLocation,
  getOpenCageLocation,
].map(
  (fn) =>
    (...params) =>
      fn(...params, axios)
);

export default (app, redis, geoCodeFunctions) => {
  app.get("/geocode", async (req, res) => {
    console.time("geocode");
    const { address } = req.query;

    const oldValue = await redis.getAsync(address);
    if (oldValue) {
      console.timeEnd("geocode");
      res.json(JSON.parse(oldValue));
    } else {
      try {
        const results = (
          await Promise.allSettled(
            geoCodeFunctions.map((func) => func(address))
          )
        )
          .filter(({ status }) => status === "fulfilled")
          .map(({ value }) => value);
        const sums = results.reduce((sums, result) => ({
          lng: sums.lng + result.lng,
          lat: sums.lat + result.lat,
        }));
        const averages = Object.fromEntries(
          Object.entries(sums).map(([key, value]) => [
            key,
            value / results.length,
          ])
        );

        redis.set(address, JSON.stringify(averages), "EX", 30);
        console.timeEnd("geocode");
        res.json(averages);
      } catch (error) {
        console.error(error);

        res.end();
      }
    }
  });
};
