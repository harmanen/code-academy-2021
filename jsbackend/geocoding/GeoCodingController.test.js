import { jest } from "@jest/globals";

import GeoCodingController, {
  getMapBoxLocation,
  getHereLocation,
  getOpenCageLocation,
} from "./GeoCodingController.js";

const redisDb = {
  Kuopio: '{"lat": 27.678281, "lng": 62.897701}',
};
const geocodeDb = {
  Helsinki: { lat: 27.678281, lng: 62.897701 },
};

let callEndpoint, mockRedis, mockGeocodeFunction;

beforeEach(() => {
  const mockApp = {
    get: jest.fn(),
  };
  mockRedis = {
    getAsync: jest.fn((address) => Promise.resolve(redisDb[address])),
    set: jest.fn(),
  };
  mockGeocodeFunction = jest.fn((address) =>
    Promise.resolve(geocodeDb[address])
  );
  GeoCodingController(mockApp, mockRedis, [mockGeocodeFunction]);

  expect(mockApp.get.mock.calls.length).toBe(1);
  const path = mockApp.get.mock.calls[0][0];
  expect(path).toBe("/geocode");
  callEndpoint = mockApp.get.mock.calls[0][1];
});

test("Test case where value is in redis", async () => {
  const end = jest.fn();
  const json = jest.fn();
  await callEndpoint({ query: { address: "Kuopio" } }, { end, json });

  expect(mockRedis.getAsync.mock.calls.length).toBe(1);
  expect(mockRedis.getAsync.mock.calls[0][0]).toBe("Kuopio");

  expect(json.mock.calls.length).toBe(1);
  expect(json.mock.calls[0][0]).toMatchSnapshot();
});

test("Test case where value is not in redis", async () => {
  const end = jest.fn();
  const json = jest.fn();
  await callEndpoint({ query: { address: "Helsinki" } }, { end, json });

  expect(mockRedis.getAsync.mock.calls.length).toBe(1);
  expect(mockRedis.getAsync.mock.calls[0][0]).toBe("Helsinki");

  expect(mockGeocodeFunction.mock.calls.length).toBe(1);
  expect(mockGeocodeFunction.mock.calls[0][0]).toBe("Helsinki");

  expect(mockRedis.set.mock.calls.length).toBe(1);
  expect(mockRedis.set.mock.calls[0][0]).toBe("Helsinki");

  expect(json.mock.calls.length).toBe(1);
  expect(json.mock.calls[0][0]).toMatchSnapshot();
});

test("Test mapbox url", async () => {
  getMapBoxLocation("Address 123, Country", {
    get: (...params) => {
      expect(params).toStrictEqual([
        `https://api.mapbox.com/geocoding/v5/mapbox.places/Address 123, Country.json?access_token=${process.env.MAPBOX_API_KEY}`,
      ]);
    },
  });
});
