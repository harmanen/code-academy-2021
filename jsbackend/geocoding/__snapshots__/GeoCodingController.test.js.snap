// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test case where value is in redis 1`] = `
Object {
  "lat": 27.678281,
  "lng": 62.897701,
}
`;

exports[`Test case where value is not in redis 1`] = `
Object {
  "lat": 27.678281,
  "lng": 62.897701,
}
`;
